enso/experiment/crfac.py

```diff
@@ -0,0 +1,70 @@
+"""Module for any LR-style experiment."""
+from sklearn.linear_model import LogisticRegression
+import sklearn_crfsuite
+import numpy as np
+import pandas as pd
+import spacy
+import crf_processing
+from enso.experiment import ClassificationExperiment
+from enso.experiment.grid_search import GridSearch
+from enso.registry import Registry, ModeKeys
+
+
+@Registry.register_experiment(
+    ModeKeys.RATIONALIZED, requirements=[("Featurizer", "PlainTextFeaturizer")]
+)
+class CRFLogit(ClassificationExperiment):
+    """Implementation of a grid-search optimized Logistic Regression model."""
+    NLP = None
+
+    def __init__(self, *args, **kwargs):
+        """Initialize internal classifier."""
+        super().__init__(auto_resample=False, *args, **kwargs)
+
+        if self.NLP is None:
+            self.NLP = spacy.load('en_vectors_web_lg')
+        self.model = LogisticRegression()
+        # Params currently fixed. TODO: sweep
+        self.extraction_model = sklearn_crfsuite.CRF(
+                algorithm='lbfgs',
+                c1=0.1,
+                c2=0.1,
+                max_iterations=100,
+                all_possible_transitions=True
+        )
+
+        self.param_grid = {
+            "penalty": ["l2"],
+            "max_iter": [500],
+            "C": [0.1, 1.0, 10.0, 100.0, 1000.0],
+            "solver": ["lbfgs"],
+            "multi_class": ["multinomial"],
+        }
+    def spacy_vectorize(self, texts):
+        return [i.vector for i in self.NLP.pipe(texts, disable=['parser', 'ner', 'textcat'])]
+
+
+    def fit(self, X, Y):
+
+        # Make special features for the CRF model
+        spans = crf_processing.get_spans_enso(X, Y)
+        spantexts = [crf_processing.get_span_texts(span) for span in spans]
```

Comment:
(just making notes to follow the code through) span text is a list of tuples `(text, "0" or "rationale")`